### ISSUE TITLE
Misplaced semicolon not detected

### ISSUE THREAD

**Issue by Zash** [mpeterv/luacheck#44]
luacheck does not detect this misplaced semicolon which casues a parse error under Lua 5.1:

``` lua
if somecondition then; -- see that semicolon?
    do some_stuff() end
end
```

This is however legal syntax in Lua 5.2, but I think it should at least warn about the empty statement.

**Comment by mpeterv** [comment]
I agree. Looks like Lua 5.1 does allow semicolons at the end of all statements, even blocks, like `do end;`, but I think Luacheck should warn about any semicolons not following an expression.

**Comment by mpeterv** [comment]
Implemented on master branch, let me know if it works for you. I'll release a new version today or tomorrow.

**Comment by Zash** [comment]
Nice, it seems to detect cases that would upset Lua 5.1.

It now warns about semicolon after `break`, which should probably be treated as zero item `return`?

```lua
for x in y do
  if x then
    break;
  end
end
```

I find it a bit unintuitive that

``` lua
local function foo() end;
```

is treated differently from

``` lua
local foo = function() end;
```

or maybe it makes sense?

**Comment by Zash** [comment]
Button too close

**Comment by mpeterv** [comment]
Right, I forgot `break`, I'll add it to exception list.

>I find it a bit unintuitive that
> `local function foo() end;`
>is treated differently from
>`local foo = function() end;`
>or maybe it makes sense?

The latter ends with an expression so it can be continued, the former one ends with a block. Do you think both this cases should be prohibited or allowed?

**Comment by mpeterv** [comment]
Some options:

1. Warn only about semicolons that are syntax errors in Lua 5.1, that is, allow semicolons after all statements, including blocks like `do ... end;`.
2. Allow semicolons only after an expression (semicolons are only needed between a statement ending with an expression and a statement starting with `(`).
3. Allow semicolons according to some style preferences (after single-line statements? Between two statements on the same line?)

**Comment by mpeterv** [comment]
Option 1 seems like the least restrictive.
